tlvc/arch/i86/drivers/block/directhd.c

```diff
@@ -678,7 +597,8 @@ void do_directhd_request(void)
 
 	if (directhd_initialized != 1) {
 	    end_request(0);
-	    continue;
+	    //continue;		/* no reason to continue */
+	    return;
```

Comment:
As per my comment in https://github.com/Mellvik/TLVC/pull/95#issuecomment-2430080543, this is a good example of an illegal action in an async I/O interrupt handler. The problem here is that if execution returns (instead of continues) here, and the request queue is non-empty, nothing will ever restart any queued I/O request, and add_request will only continue to queue I/O, never calling the request function directly as it does when the queue is empty. The driver will effective hang and never continue any existing nor future I/O request, even if a later open succeeds.

In the above case, the system will just hang by sleeping all processes waiting for I/O on this device, permanently when the I/O queue max size is reached, with no indication of why the failure is occurring. Continuing would at least empty the queue, so that the I/O request subsystem continues to function. Of course, all this begs the question of how/why an I/O request can be submitted when the driver open routine failed. IMO it can't. So that begs the question of why check for driver initialization in the request routine. IMO there's no point in it, that's what the _open routine return value is for, and the buffer system will never submit a block I/O request on an unopened device. So it seems the most correct action is to discard the initialization check and don't return or continue!  If the driver can't complete the request for any reason, a later call to end_request(0) I'm sure is already coded somewhere.


FYI.

Note that the current DF driver has no such initialized check.

Reply:
Thanks for the code review and the comment @ghaerr. This is not an async driver - and turning it into one is quite a bit down on the list because there is minimal gain, possibly the opposite because everything it PIO and consuming cycles anyway.